backends/ubpf/ubpfRegister.cpp

```diff
@@ -0,0 +1,174 @@
+            return;
+        } else if (method->method->name.name ==
+                   program->model.registerModel.write.name) {
+            emitRegisterWrite(builder, method->expr, pointerVariables);
+            return;
+        }
+        error("%1%: Unexpected method for %2%", method->expr,
+              program->model.registerModel.read.name);
+    }
+
+    void UBPFRegister::emitRegisterWrite(EBPF::CodeBuilder *builder,
+                                         const IR::MethodCallExpression *expression,
+                                         const std::vector<cstring> &pointerVariables) {
+        BUG_CHECK(expression->arguments->size() == 2,
+                  "Expected just 2 argument for %1%", expression);
+
+        auto arg_value = expression->arguments->at(1);
+        auto target = (UbpfTarget *) builder->target;
+
+        cstring valueVariableName = nullptr;
+        valueVariableName = emitValueInstanceIfNeeded(builder, arg_value);
+
+        if (arg_value->expression->is<IR::PathExpression>()) {
+            auto name = arg_value->expression->to<IR::PathExpression>()->path->name.name;
+            if (std::find(
+                    pointerVariables.begin(),
+                    pointerVariables.end(),
+                    name) ==
+                pointerVariables.end()) {
+
+                valueVariableName = name;
+            }
+        }
+
+        if (valueVariableName != nullptr) {
+            target->emitTableUpdate(builder, dataMapName, last_key_name,
+                                    "&" + valueVariableName);
+        } else {
+            target->emitTableUpdate(codeGen, builder, dataMapName, last_key_name,
+                                    arg_value->expression);
+        }
+    }
+
+    cstring UBPFRegister::emitValueInstanceIfNeeded(EBPF::CodeBuilder *builder,
+                                                    const IR::Argument *arg_value) {
+
+        cstring valueVariableName = nullptr;
+
+        if (arg_value->expression->is<IR::Constant>() ||
+            arg_value->expression->is<IR::Operation_Binary>() ||
+            arg_value->expression->is<IR::Member>()) {
+
+            auto scalarInstance = UBPFTypeFactory::instance->create(
+                    valueType);
+            valueVariableName = program->refMap->newName(
+                    "value_local_var");
+            scalarInstance->declare(builder, valueVariableName, false);
+            builder->append(" = ");
+            codeGen->visit(arg_value->expression);
+            builder->endOfStatement(true);
+            builder->emitIndent();
+        }
+
+        return valueVariableName;
+    }
+
+    void UBPFRegister::emitRegisterRead(EBPF::CodeBuilder *builder,
+                                        const IR::MethodCallExpression *expression) {
+        BUG_CHECK(expression->arguments->size() == 1,
+                  "Expected 1 argument for %1%", expression);
+        auto target = (UbpfTarget *) builder->target;
```

Comment:
Do you need this cast? Wouldn't virtual dispatch do the right thing anyway?

Reply:
Removed.